src/InformaticsGateway/Services/Scp/ApplicationEntityHandler.cs

```diff
@@ -0,0 +1,103 @@
+{
+    internal class ApplicationEntityHandler : IDisposable
+    {
+        private readonly object _syncRoot = new object();
+        private bool _disposed = false;
+        private IServiceScopeFactory _serviceScopeFactory;
+        private MonaiApplicationEntity _configuration;
+        private IStorageInfoProvider _storageInfoProvider;
+        private IPayloadAssembler _payloadAssembler;
+        private IFileSystem _fileSystem;
+        private IDicomToolkit _dicomToolkit;
+        private ILogger<ApplicationEntityHandler> _logger;
+
+        public ApplicationEntityHandler(
+            IServiceScopeFactory serviceScopeFactory,
+            MonaiApplicationEntity monaiApplicationEntity,
+            IStorageInfoProvider storageInfoProvider,
+            IPayloadAssembler payloadAssembler,
+            IFileSystem fileSystem,
+            IDicomToolkit dicomToolkit,
+            ILogger<ApplicationEntityHandler> logger)
+        {
+            _serviceScopeFactory = serviceScopeFactory ?? throw new ArgumentNullException(nameof(serviceScopeFactory));
+            _configuration = monaiApplicationEntity ?? throw new ArgumentNullException(nameof(monaiApplicationEntity));
+            _storageInfoProvider = storageInfoProvider ?? throw new ArgumentNullException(nameof(storageInfoProvider));
+            _payloadAssembler = payloadAssembler ?? throw new ArgumentNullException(nameof(payloadAssembler));
+            _fileSystem = fileSystem ?? throw new ArgumentNullException(nameof(fileSystem));
+            _dicomToolkit = dicomToolkit ?? throw new ArgumentNullException(nameof(dicomToolkit));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        ~ApplicationEntityHandler() => Dispose(false);
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (!_disposed)
+            {
+                if (disposing)
+                {
+                }
+                _disposed = true;
+            }
+        }
+
+        internal async Task HandleInstance(DicomCStoreRequest request, DicomFileStorageInfo info)
+        {
+            if (_configuration.IgnoredSopClasses.Contains(request.SOPClassUID.UID))
+            {
+                _logger.Log(LogLevel.Information, $"Instance ignored due to matching SOP Class UID {request.SOPClassUID.UID}");
+                return;
+            }
+
+            if (_configuration.Workflows.Any())
+            {
+                info.SetWorkflows(_configuration.Workflows.ToArray());
+            }
+
+            await SaveDicomInstance(request, info.FilePath);
+
+            var dicomTag = FellowOakDicom.DicomTag.Parse(_configuration.Grouping);
+            if (request.Dataset.TryGetSingleValue<string>(dicomTag, out string key))
+            {
+                _payloadAssembler.Queue(key, info, _configuration.Timeout);
```

Comment:
I don't see testing the value of the key and handling the case where the expected tag is missing. I was looking for constraints on the Grouping tags, but did not see it except for the validity check of the tags, so is it true the grouping tag can be any valid one? If somehow the grouping tag is one of the SQ VR type, what is expected to happen then?

Reply:
Testing of the key is done in the configuration validation class and is done during startup. Failures to provide a valid DICOM tag results in application crash.

Reply:
Perhaps we shall just limit to the 3 options: patient, study, and series.

Reply:
Always being my position to only allow tags that make sense clinically and logically, not just because technically one can try to group instances with any of the tags.

Another important tag often required/used in the radiology workflow is the accession number of a specific issuer, but it is a type 2 tag and may be missing.

For grouping with Patient ID (not `patient`), again, the Issuer of Patient ID needs to be included to scope the patient ID. Also, they are type 2 and 3 tags.

Reply:
Limiting to Study Instance UID (0020,000D) & Series Instance UID (0020,000E) for now; updated docs